File: benchmarking/sequence_algorithms.py
from __future__ import print_function, division
import numpy as np
from Bio import Phylo


if __name__ == '__main__':
    from treetime.seq_utils import normalize_profile, prof2seq, seq2prof
    from treetime.gtr import GTR    

    gtr = GTR.standard('JC69')
    dummy_prof = np.random.random(size=(10000,5))

    # used a lot (300us)
    norm_prof = normalize_profile(dummy_prof)[0]

    # used less but still a lot (50us)
    gtr.evolve(norm_prof, 0.1)

    # used less but still a lot (50us)
    gtr.propagate_profile(norm_prof, 0.1)

    # used only in final, sample_from_prof=False speeds it up (600us or 300us)
    seq, p, seq_ii = prof2seq(norm_prof, gtr, sample_from_prof=True, normalize=False)

    # used only initially (slow, 5ms)
    tmp_prof = seq2prof(seq, gtr.profile_map)

File: packages/treetime/examples/sequence_algorithms.rs
use ctor::ctor;
use eyre::Report;
use ndarray::Array2;
use ndarray_rand::RandomExt;
use treetime::nuc_models::jc69::jc69;
use treetime::seq_utils::normalize_profile::normalize_profile;
use treetime::seq_utils::seq2prof::{prof2seq, seq2prof};
use treetime::utils::global_init::global_init;

#[cfg(all(target_family = "linux", target_arch = "x86_64"))]
#[global_allocator]
static GLOBAL: mimalloc::MiMalloc = mimalloc::MiMalloc;

#[ctor]
fn init() {
  global_init();
}

fn main() -> Result<(), Report> {
  let gtr = jc69()?;

  let dummy_prof = Array2::<f32>::random((10000, 5));

  // used a lot (300us)
  let (norm_prof, _) = normalize_profile(&dummy_prof, false);

  // used less but still a lot (50us)
  gtr.evolve(&norm_prof, 0.1, false);

  // used less but still a lot (50us)
  gtr.propagate_profile(norm_prof, 0.1, false);

  // // used only in final, sample_from_prof=False speeds it up (600us or 300us)
  // let  (seq, p, seq_ii) = prof2seq(norm_prof, &gtr, true, false);
  //
  // // used only initially (slow, 5ms)
  // let tmp_prof = seq2prof(seq, gtr.profile_map);

  Ok(())
}
